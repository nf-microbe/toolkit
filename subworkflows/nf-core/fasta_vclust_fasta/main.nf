// Import modules
include { SEQHASHER_SEQUNIQ } from '../../../modules/nf-core/seqhasher/sequniq'
include { SEQKIT_GREP       } from '../../../modules/nf-core/seqkit/grep'
include { VCLUST_ALIGN      } from '../../../modules/nf-core/vclust/align'
include { VCLUST_CLUSTER    } from '../../../modules/nf-core/vclust/cluster'
include { VCLUST_PREFILTER  } from '../../../modules/nf-core/vclust/prefilter'

workflow FASTA_VCLUST_FASTA {
    take:
    fasta_gz    // [ [ meta ], fasta.gz ]
    vclust      // path/to/vclust/binary

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Prefilter alignments
    //
    VCLUST_PREFILTER(
        fasta_gz,
        vclust
    )
    ch_versions = ch_versions.mix(VCLUST_PREFILTER.out.versions)

    //
    // MODULE: Align sequences
    //
    VCLUST_ALIGN(
        fasta_gz,
        VCLUST_PREFILTER.out.prefilter,
        vclust
    )
    ch_versions = ch_versions.mix(VCLUST_ALIGN.out.versions)

    //
    // MODULE: Cluster sequences
    //
    VCLUST_CLUSTER(
        VCLUST_ALIGN.out.ani,
        VCLUST_ALIGN.out.ids,
        vclust
    )
    ch_versions = ch_versions.mix(VCLUST_CLUSTER.out.versions)

    //
    // MODULE: Dereplicate cluster output to retain only cluster rep
    //
    SEQHASHER_SEQUNIQ(
        VCLUST_CLUSTER.out.clusters
    )
    ch_versions = ch_versions.mix(SEQHASHER_SEQUNIQ.out.versions)

    //
    // MODULE: Extract rep sequences from FastA file
    //
    SEQKIT_GREP(
        fasta_gz,
        SEQHASHER_SEQUNIQ.out.unique.map { meta, tsv -> tsv }
    )
    ch_versions = ch_versions.mix(SEQKIT_GREP.out.versions)

    emit:
    cluster_reps_fasta_gz   = SEQKIT_GREP.out.filter    // [ [ meta ], fasta.gz ]
    versions                = ch_versions               // channel: [ versions.yml ]
}
