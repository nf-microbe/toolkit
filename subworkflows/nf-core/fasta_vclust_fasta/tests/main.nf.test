nextflow_workflow {

    name "Test Subworkflow FASTA_VCLUST_FASTA"
    script "../main.nf"
    workflow "FASTA_VCLUST_FASTA"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fasta_vclust_fasta"
    tag "seqhasher"
    tag "seqhasher/sequniq"
    tag "seqkit"
    tag "seqkit/grep"
    tag "vclust"
    tag "vclust/align"
    tag "vclust/cluster"
    tag "vclust/prefilter"
    tag "cat/cat"

    setup {
        run("CAT_CAT") {
            script "../../../../modules/nf-core/cat/cat/main.nf"
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true),
                        file("${projectDir}/assets/test-datasets/genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test2.contigs.fa.gz", checkIfExists: true)
                    ]
                ]
                """
            }
        }
    }

    test("b_fragilis - fasta.gz") {
        when {
            params {
                vclust_cluster_args = "--algorithm leiden --metric ani --ani 0.95 --qcov 0.85 --out-repr"
            }
            workflow {
                """
                input[0] = CAT_CAT.out.file_out
                input[1] = file("${projectDir}/bin/vclust", checkIfExists: true)
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
