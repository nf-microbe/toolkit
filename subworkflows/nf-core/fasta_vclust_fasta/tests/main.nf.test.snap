{
    "b_fragilis - fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cluster_reps.fa.gz:md5,3dcef382a0f144a7d4282649cc3e0d0c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,4152e9af05390f2ccffe6cbad417aa5a",
                    "versions.yml:md5,4e1b518e8c3d0bfb49020fa3b6fc9b8b",
                    "versions.yml:md5,795c024849c14da1c9f587bc2d07aa6e",
                    "versions.yml:md5,9b8eb4e71f4dca0004ed0377e39b965e",
                    "versions.yml:md5,9fd2d2cb61702b996ba6895786064cb1"
                ],
                "cluster_reps_fasta_gz": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cluster_reps.fa.gz:md5,3dcef382a0f144a7d4282649cc3e0d0c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4152e9af05390f2ccffe6cbad417aa5a",
                    "versions.yml:md5,4e1b518e8c3d0bfb49020fa3b6fc9b8b",
                    "versions.yml:md5,795c024849c14da1c9f587bc2d07aa6e",
                    "versions.yml:md5,9b8eb4e71f4dca0004ed0377e39b965e",
                    "versions.yml:md5,9fd2d2cb61702b996ba6895786064cb1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-02T13:45:28.469004246"
    }
}